upb/decode_fast.c

```diff
@@ -0,0 +1,330 @@
+    return fastdecode_varint(UPB_PARSE_ARGS, tagbytes, valbytes, CARD_##card,  \
+                             type##_ZZ);                                       \
+  }
+
+#define TYPES(card, tagbytes) \
+  F(card, b, 1, tagbytes)     \
+  F(card, v, 4, tagbytes)     \
+  F(card, v, 8, tagbytes)     \
+  F(card, z, 4, tagbytes)     \
+  F(card, z, 8, tagbytes)
+
+#define TAGBYTES(card) \
+  TYPES(card, 1)       \
+  TYPES(card, 2)
+
+TAGBYTES(s)
+TAGBYTES(o)
+//TAGBYTES(r)
+
+#undef z_ZZ
+#undef b_ZZ
+#undef v_ZZ
+#undef o_ONEOF
+#undef s_ONEOF
+#undef r_ONEOF
+#undef F
+#undef TYPES
+#undef TAGBYTES
+
+/* string fields **************************************************************/
+
+UPB_FORCEINLINE
+bool fastdecode_boundscheck(const char *ptr, unsigned len, const char *end) {
+  uintptr_t uptr = (uintptr_t)ptr;
+  uintptr_t uend = (uintptr_t)end;
+  uintptr_t res = uptr + len;
+  return res < uptr || res > uend;
+}
+
+UPB_FORCEINLINE
+static const char *fastdecode_string(UPB_PARSE_PARAMS, int tagbytes,
+                                     upb_card card) {
+  upb_strview *dst;
+  int64_t len;
+  if (UPB_UNLIKELY(!fastdecode_checktag(data, tagbytes))) {
+    return fastdecode_generic(UPB_PARSE_ARGS);
+  }
+
+  dst = fastdecode_getfield(d, ptr, msg, &data, &hasbits, tagbytes,
+                            sizeof(upb_strview), card);
+  len = ptr[tagbytes];
+  if (UPB_UNLIKELY(len < 0)) {
+    return fastdecode_generic(UPB_PARSE_ARGS);
+  }
+  ptr += tagbytes + 1;
+  dst->data = ptr;
+  dst->size = len;
+  if (UPB_UNLIKELY(fastdecode_boundscheck(ptr, len, d->limit))) {
+    return fastdecode_err(d);
+  }
+  ptr += len;
+  return fastdecode_dispatch(d, ptr, msg, table, hasbits);
+}
+
+const char *upb_pss_1bt(UPB_PARSE_PARAMS) {
+  return fastdecode_string(UPB_PARSE_ARGS, 1, CARD_s);
+}
+
+const char *upb_pos_1bt(UPB_PARSE_PARAMS) {
+  return fastdecode_string(UPB_PARSE_ARGS, 1, CARD_o);
+}
+
+const char *upb_pss_2bt(UPB_PARSE_PARAMS) {
+  return fastdecode_string(UPB_PARSE_ARGS, 2, CARD_s);
+}
+
+const char *upb_pos_2bt(UPB_PARSE_PARAMS) {
+  return fastdecode_string(UPB_PARSE_ARGS, 2, CARD_o);
+}
+
+/* message fields *************************************************************/
+
+UPB_FORCEINLINE
+static const char *fastdecode_submsg(UPB_PARSE_PARAMS, int tagbytes,
+                                     upb_card card) {
+  const char *saved_limit;
+  const upb_msglayout_field *field = &table->fields[data >> 48];
+  size_t ofs = field->offset;
+  const upb_msglayout *subl = table->submsgs[field->submsg_index];
+  upb_array *arr;
+  upb_msg **submsg;
+
+  if (UPB_UNLIKELY(!fastdecode_checktag(data, tagbytes))) {
+    return fastdecode_generic(UPB_PARSE_ARGS);
+  }
+
+  submsg = fastdecode_getfield_ofs(d, ptr, msg, ofs, &data, &hasbits, &arr,
+                                   tagbytes, sizeof(upb_msg *), card);
+
+again:
+  if (card == CARD_r) {
+    if (UPB_UNLIKELY((uint32_t)data == 0)) {
```

Comment:
if you store data = ((elem_avail - 1) << 16) | tag
then this can be if (int64)data < 0 return fastdecode_generic

and below data -= 65536;
checking tag is than trivial byte or word compare.

Reply:
Now I am using an explicit end pointer. lmk if you think it's better to change.

Reply:
Yes I think this is good